lib/posix/pthread.c

```diff
@@ -226,22 +377,27 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
 int pthread_setcancelstate(int state, int *oldstate)
 {
 	bool cancel_pending;
-	k_spinlock_key_t cancel_key;
-	struct posix_thread *pthread = to_posix_thread(pthread_self());
+	k_spinlock_key_t key;
+	struct posix_thread *t;
 
 	if (state != PTHREAD_CANCEL_ENABLE &&
 	    state != PTHREAD_CANCEL_DISABLE) {
 		return EINVAL;
 	}
 
-	cancel_key = k_spin_lock(&pthread->cancel_lock);
-	*oldstate = pthread->cancel_state;
-	pthread->cancel_state = state;
-	cancel_pending = pthread->cancel_pending;
-	k_spin_unlock(&pthread->cancel_lock, cancel_key);
+	t = to_posix_thread(pthread_self());
+	if (t == NULL) {
+		return EINVAL;
+	}
+
+	key = k_spin_lock(&t->lock);
+	*oldstate = t->cancel_state;
+	t->cancel_state = state;
+	cancel_pending = t->cancel_pending;
+	k_spin_unlock(&t->lock, key);
 
 	if (state == PTHREAD_CANCEL_ENABLE && cancel_pending) {
-		pthread_exit((void *)PTHREAD_CANCELED);
```

Comment:
From my reading of the POSIX spec, `PTHREAD_CANCELED` should already be a `void *` value, so no cast should be needed here. Might do some checking to make sure this is true -- otherwise, you wouldn't be able to test for this value with a simple compare operation like `res == PTHREAD_CANCELED` as the POSIX examples show.

Reply:
That is true! I'll remove the case 👍